### PR TITLE
Track finalized block root index and filter in blocks by range

### LINKED ISSUE #3819: Beacon chain recurring error / warning message every 10-15 minutes
The following error / warning message recurs every 10-15 minutes:

https://cdn.discordapp.com/attachments/560308336248225812/636008981994799114/unknown.png

Copying text as error:

[2019-10-22 00:08:02] ERROR sync: Failed to handle p2p pubsub error=pre state of target block 62424 does not exist
could not process attestation from fork choice service
github.com/prysmaticlabs/prysm/beacon-chain/blockchain.(*Service).ReceiveAttestationNoPubsub
        beacon-chain/blockchain/receive_attestation.go:69
github.com/prysmaticlabs/prysm/beacon-chain/sync.(*RegularSync).beaconAttestationSubscriber
        beacon-chain/sync/subscriber_beacon_attestation.go:17
github.com/prysmaticlabs/prysm/beacon-chain/sync.(*RegularSync).subscribe.func1
        beacon-chain/sync/subscriber.go:146
runtime.goexit
        GOROOT/src/runtime/asm_amd64.s:1357 topic=/eth2/beacon_attestation/ssz

@terencechain @nisdas

### LINKED ISSUE #3848: Missing block doesn't link back to finalized check point in state
Beacon chain started receiving the following errors:  

https://cdn.discordapp.com/attachments/560308336248225812/636916326299861022/unknown.png

As text:

[2019-10-24 19:47:11] ERROR sync: Could not process block from slot 97362: could not process block from fork choice service: block from slot 97362 is not a descendent of the current finalized block
[2019-10-24 19:47:11]  INFO sync: Processed ancestor block with slot 97362 and cleared pending block cache
[2019-10-24 19:47:11]  INFO sync: Requesting parent block currentSlot=97365 parentRoot=72ce5e2a3f0acc078b3be672cf66cec29c623a85f6dcd9a94bed087c5aaad4c5
[2019-10-24 19:47:13]  INFO sync: Requesting parent block currentSlot=97365 parentRoot=72ce5e2a3f0acc078b3be672cf66cec29c623a85f6dcd9a94bed087c5aaad4c5
[2019-10-24 19:47:14]  INFO p2p: New peer connected peer=16Uiu2HAmU1MXD5wLZQ3dVgMnuiad8c2y39X9bH7rwH8w953J4JAK
[2019-10-24 19:47:15]  INFO sync: Requesting parent block currentSlot=97365 parentRoot=72ce5e2a3f0acc078b3be672cf66cec29c623a85f6dcd9a94bed087c5aaad4c5
[2019-10-24 19:47:17] ERROR sync: Could not process block from slot 97362: could not process block from fork choice service: block from slot 97362 is not a descendent of the current finalized block
[2019-10-24 19:47:17]  INFO sync: Processed ancestor block with slot 97362 and cleared pending block cache
[2019-10-24 19:47:17]  INFO sync: Requesting parent block currentSlot=97365 parentRoot=72ce5e2a3f0acc078b3be672cf66cec29c623a85f6dcd9a94bed087c5aaad4c5
[2019-10-24 19:47:19]  INFO sync: Requesting parent block currentSlot=97365 parentRoot=72ce5e2a3f0acc078b3be672cf66cec29c623a85f6dcd9a94bed087c5aaad4c5

Per @terencechain : "Looks like you got peered with a node that gave you a missing block that doesn't link back to your finalized check point in state. Regardless we should have a way to exit such loop."

### LINKED ISSUE #3858: Initial sync stuck at slot 101323. Peers serving non-canonical block
In the current running testnet, block at slot 101323 has a parent that is older than a finalized epoch.

101285 is the parent of block at slot 101323
101288 is finalized
syncing continues until 101323 then we see the following error
```
[2019-10-26 14:07:23]  INFO initial-sync: Processing block 101323/132617 - estimated time remaining 5h25m58s blocksPerSecond=1.6 peers=1/8
panic: could not process block from fork choice service: block from slot 101323 is not a descendent of the current finalized block

goroutine 260 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*InitialSync).Start(0xc0007abe90)
        beacon-chain/sync/initial-sync/service.go:106 +0x97d
created by github.com/prysmaticlabs/prysm/shared.(*ServiceRegistry).StartAll
        shared/service_registry.go:44 +0x23e
```

### PR DESCRIPTION
**Pull request by prestonvanloon**
Resolves #3858. 
Resolves #3819.
Resolves #3848.

This PR keeps track of finalized block roots in the database. This index can then be queried to ensure that the block is a finalized block. When serving blocks at RPC, the beacon chain node will query the index to filter non-finalized blocks.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
I want to handle the case where the index is being built, but the process is interrupted. This could lead to an invalid / incomplete index.

File: beacon-chain/db/kv/finalized_block_roots.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+			}
+			block := &ethpb.BeaconBlock{}
+			if err := proto.Unmarshal(enc, block); err != nil {
+				traceutil.AnnotateError(span, err)
+				return err
+			}
+
+			container := &dbpb.FinalizedBlockRootContainer{
+				ParentRoot: block.ParentRoot,
+				ChildRoot:  previousRoot,
```
Comment (nisdas) [inline beacon-chain/db/kv/finalized_block_roots.go]: The first time the walk is performed , `previousRoot` will be an empty byte slice

File: beacon-chain/db/kv/finalized_block_roots.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+	root := checkpoint.Root
+	var previousRoot []byte
+	genesisRoot := blocks.Get(genesisBlockRootKey)
+
+	// Walk up the ancestry chain until we reach a block root present in the finalized block roots
+	// index bucket or genesis block root.
+	var walk = func() error {
+		for {
+			if bytes.Equal(root, genesisRoot) {
```
Comment (nisdas) [inline beacon-chain/db/kv/finalized_block_roots.go]: this walks all the way back to genesis each time we updated the finalized checkpoint, shouldn't it only walk back to the previous finalized checkpoint.

File: beacon-chain/db/kv/finalized_block_roots.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/finalized_block_roots.go]: That is expected, there is no finalized child root yet. We just learned of a new finalized block and this is the highest finalized block to exist so far.

File: beacon-chain/db/kv/finalized_block_roots.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/db/kv/finalized_block_roots.go]: It does, see line 67

File: validator/BUILD.bazel [hunk 3]
```diff
@@ -77,7 +77,7 @@ docker_push(
 go_binary(
     name = "validator",
     embed = [":go_default_library"],
-    pure = "on",  # Enabled unless there is a valid reason to include cgo dep.
+    pure = "off",  # Enabled unless there is a valid reason to include cgo dep.
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply validator/BUILD.bazel]: oops, will revert these

File: beacon-chain/db/kv/finalized_block_roots.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+					traceutil.AnnotateError(span, err)
+					return err
+				}
+				parent.ChildRoot = root
+				enc, err := proto.Marshal(parent)
+				if err != nil {
+					traceutil.AnnotateError(span, err)
+					return err
+				}
+				return bkt.Put(block.ParentRoot, enc)
```
Comment (nisdas) [inline beacon-chain/db/kv/finalized_block_roots.go]: ah sorry just saw this, its good then in that case

File: beacon-chain/db/kv/finalized_block_roots_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+var genesisBlockRoot = bytesutil.ToBytes32([]byte{'G', 'E', 'N', 'E', 'S', 'I', 'S'})
+
+func init() {
+	fc := featureconfig.Get()
+	fc.EnableFinalizedBlockRootIndex = true
+	featureconfig.Init(fc)
+}
+
+func TestStore_IsFinalizedBlock(t *testing.T) {
```
Comment (nisdas) [inline beacon-chain/db/kv/finalized_block_roots_test.go]: Is it possible to cover a few more cases ? ex: Across multiple epochs, with skipped slots at epoch boundaries,etc

Comment (prestonvanloon) [reply beacon-chain/db/kv/finalized_block_roots_test.go]: Hmmm, I'll think about overnight. There isn't any different code path for mulitple epochs with skipped slots. Epoch boundries aren't relevant here. As long as the finalized chain is a chain then this should work.

**Pull request merged**